src/layout/rect.rs

```diff
@@ -86,6 +116,75 @@ impl Rect {
         }
     }
 
+    /// Adjusts the size of the `Rect`.
+    ///
+    /// Adjust the size of the `Rect` by moving either the top or left. This effectively changes
+    /// [`Rect::x`] and [`Rect::y`] and recalculate [`Rect::width`] and [`Rect::height`] so that
+    /// the right and bottom "borders" do not move.
```

Comment:
It is an odd behavior indeed.

The original goal was to avoid those kind of computation that are common:

[calendar.rs](https://github.com/ratatui-org/ratatui/blob/c862aa5e9ef4dbf494b5151214ac87f5c71e76d4/examples/calendar.rs#L41)
![image](https://github.com/ratatui-org/ratatui/assets/36198422/84ddc90a-6148-475d-8bad-8a29ff2b08ef)
[widgets/barchart.rs](https://github.com/ratatui-org/ratatui/blob/c862aa5e9ef4dbf494b5151214ac87f5c71e76d4/src/widgets/barchart.rs#L394)
![image](https://github.com/ratatui-org/ratatui/assets/36198422/dd8b8424-719e-4d78-82e3-9aa0cc28b508)

But looking back at it it seems to complex.

We probably shouldn't make very specific methods and instead more general ones as you proposed earlier in this review.